File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.java
/*
 * Copyright (c) 2021 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package org.forgerock.android.auth.callback;

import static androidx.biometric.BiometricManager.Authenticators.BIOMETRIC_STRONG;
import static androidx.biometric.BiometricManager.Authenticators.DEVICE_CREDENTIAL;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Build;
import android.os.OperationCanceledException;

import androidx.annotation.Keep;
import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.biometric.BiometricPrompt;
import androidx.fragment.app.FragmentActivity;

import com.nimbusds.jose.JWSAlgorithm;
import com.nimbusds.jose.JWSHeader;
import com.nimbusds.jose.crypto.RSASSASigner;
import com.nimbusds.jwt.JWTClaimsSet;
import com.nimbusds.jwt.SignedJWT;

import org.forgerock.android.auth.FRListener;
import org.forgerock.android.auth.InitProvider;
import org.forgerock.android.auth.Listener;
import org.forgerock.android.auth.StringUtils;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.security.GeneralSecurityException;
import java.security.KeyStore;
import java.security.PrivateKey;

import lombok.Getter;
import lombok.NoArgsConstructor;

@NoArgsConstructor
@Getter
public class DeviceSigningVerifierCallback extends AbstractCallback implements KeyAware {

    private String userId;
    private String challenge;
    private String title;
    private Integer timeout;
    private String subtitle;
    private String description;

    @Keep
    public DeviceSigningVerifierCallback(JSONObject jsonObject, int index) {
        super(jsonObject, index);
    }

    @Override
    protected void setAttribute(String name, Object value) {
        switch (name) {
            case "userId":
                this.userId = (String) value;
                break;
            case "challenge":
                this.challenge = (String) value;
                break;
            case "title":
                this.title = (String) value;
                break;
            case "subtitle":
                this.subtitle = (String) value;
                break;
            case "description":
                this.description = (String) value;
                break;
            case "timeout":
                this.timeout = (Integer) value;
                break;

            default:
                //ignore
        }
    }

    public String getUserId() {
       if (StringUtils.isEmpty(userId)) {
            //Get the first one for now.
           SharedPreferences sharedPreferences = InitProvider.getCurrentActivity()
                   .getSharedPreferences(DEVICE_BINDING, Context.MODE_PRIVATE);
           if (sharedPreferences.getAll().keySet().isEmpty()) {
                throw new IllegalStateException("User Not found");
            } else {
                try {
                    return new JSONObject((String)sharedPreferences.getAll().entrySet().iterator().next().getValue()).getString("userId");
                } catch (JSONException e) {
                    throw new IllegalStateException(e);
                }
            }
        } else {
            return userId;
        }
    }

    public void setJws(String value) {
        super.setValue(value, 0);
    }

    public void setClientError(String value) {
        super.setValue(value, 1);
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    public void sign(Context context, FRListener<Void> listener) {

        //Check if the device support it.
        if (!isSupported()) {
            setClientError("unsupported");
            Listener.onException(listener, new UnsupportedOperationException());
        }

        FragmentActivity fragmentActivity = InitProvider.getCurrentActivityAsFragmentActivity();

        String keyAlias = null;
        try {
            keyAlias = getKeyAlias();
        } catch (IllegalStateException e) {
            setClientError("keyNotFound");
            Listener.onException(listener, new IllegalStateException("Key not found") );
            return;
        }
        if (getKey(keyAlias) == null)  {
            setClientError("keyNotFound");
            Listener.onException(listener, new IllegalStateException("Key not found") );
            return;
        }

        String finalKeyAlias = keyAlias;
        BiometricPrompt biometricPrompt = new BiometricPrompt(fragmentActivity, new BiometricPrompt.AuthenticationCallback() {
            @Override
            public void onAuthenticationError(int errorCode, @NonNull CharSequence errString) {
                super.onAuthenticationError(errorCode, errString);
                setClientError("unsupported");
                Listener.onException(listener, new OperationCanceledException(errString.toString()));
            }

            @Override
            public void onAuthenticationSucceeded(@NonNull BiometricPrompt.AuthenticationResult result) {
                super.onAuthenticationSucceeded(result);
                try {
                    SignedJWT signedJWT = new SignedJWT(
                            new JWSHeader.Builder(JWSAlgorithm.RS512)
                                    .keyID(getBindingInfo(context).getString("kid"))
                                    .build()
                            , new JWTClaimsSet.Builder()
                            .subject(getUserId())
                            .claim("challenge", challenge)
                            .build());

                    signedJWT.sign(new RSASSASigner(getPrivateKey(finalKeyAlias)));
                    setJws(signedJWT.serialize());
                    Listener.onSuccess(listener, null);
                } catch (Exception e) {
                    Listener.onException(listener, e);
                }
            }

            @Override
            public void onAuthenticationFailed() {
                super.onAuthenticationFailed();
                setClientError("unsupported");
                Listener.onException(listener, new OperationCanceledException());
            }

        });

        BiometricPrompt.PromptInfo promptInfo = new BiometricPrompt.PromptInfo.Builder()
                .setTitle(getTitle())
                .setSubtitle(getSubtitle())
                .setConfirmationRequired(true)
                .setDescription(getDescription())
                //.setNegativeButtonText("Cancel")
                .setConfirmationRequired(true)
                .setAllowedAuthenticators(BIOMETRIC_STRONG | DEVICE_CREDENTIAL).build();
                //.setAllowedAuthenticators(BIOMETRIC_STRONG).build();

        biometricPrompt.authenticate(promptInfo);

    }

    protected PrivateKey getPrivateKey(String keyAlias) throws GeneralSecurityException, IOException {
        KeyStore keyStore = getKeyStore();
        return (PrivateKey) keyStore.getKey(keyAlias, null);
    }

    protected boolean isSupported() {
        return true;
    }

    @Override
    public String getType() {
        return DeviceSigningVerifierCallback.class.getSimpleName();
    }

}

File: forgerock-auth/src/main/java/org/forgerock/android/auth/callback/DeviceSigningVerifierCallback.kt
/*
 * Copyright (c) 2022 - 2024 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */
package org.forgerock.android.auth.callback

import android.content.Context
import androidx.annotation.Keep
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withTimeout
import org.forgerock.android.auth.FRListener
import org.forgerock.android.auth.Listener
import org.forgerock.android.auth.devicebind.DefaultUserKeySelector
import org.forgerock.android.auth.devicebind.DeviceAuthenticator
import org.forgerock.android.auth.devicebind.DeviceBindingErrorStatus
import org.forgerock.android.auth.devicebind.DeviceBindingErrorStatus.ClientNotRegistered
import org.forgerock.android.auth.devicebind.DeviceBindingErrorStatus.Unsupported
import org.forgerock.android.auth.devicebind.DeviceBindingErrorStatus.InvalidCustomClaims
import org.forgerock.android.auth.devicebind.DeviceBindingException
import org.forgerock.android.auth.devicebind.NoKeysFound
import org.forgerock.android.auth.devicebind.Prompt
import org.forgerock.android.auth.devicebind.SingleKeyFound
import org.forgerock.android.auth.devicebind.Success
import org.forgerock.android.auth.devicebind.UserDeviceKeyService
import org.forgerock.android.auth.devicebind.UserKey
import org.forgerock.android.auth.devicebind.UserKeySelector
import org.forgerock.android.auth.devicebind.UserKeyService
import org.forgerock.android.auth.devicebind.UserKeys
import org.forgerock.android.auth.devicebind.initialize
import org.json.JSONObject


/**
 * Callback to collect the device signing information
 */
open class DeviceSigningVerifierCallback : AbstractCallback, Binding {

    @Keep
    @JvmOverloads
    constructor(jsonObject: JSONObject, index: Int) : super(jsonObject, index)

    @Keep
    @JvmOverloads
    constructor() : super()

    /**
     * The optional userId
     */
    var userId: String? = null
        private set

    /**
     * The challenge received from server
     */
    lateinit var challenge: String
        private set

    /**
     * The title to be displayed in biometric prompt
     */
    lateinit var title: String
        private set

    /**
     * The subtitle to be displayed in biometric prompt
     */
    lateinit var subtitle: String
        private set

    /**
     * The description to be displayed in biometric prompt
     */
    lateinit var description: String
        private set

    /**
     * The timeout to be to expire the biometric authentication
     */
    var timeout: Int? = null
        private set

    override fun setAttribute(name: String, value: Any) = when (name) {
        "userId" -> userId = value as? String
        "challenge" -> challenge = value as String
        "title" -> title = value as? String ?: ""
        "subtitle" -> subtitle = value as? String ?: ""
        "description" -> description = value as? String ?: ""
        "timeout" -> timeout = value as? Int
        else -> {}
    }

    override fun getType(): String {
        return "DeviceSigningVerifierCallback"
    }

    /**
     * Input the JWS key to the server
     * @param value The JWS value.
     */
    fun setJws(value: String?) {
        super.setValue(value, 0)
    }


    /**
     * Input the Client Error to the server
     * @param value DeviceSign ErrorType .
     */
    override fun setClientError(clientError: String?) {
        super.setValue(clientError, 1)
    }

    /**
     * Sign the challenge with bounded device keys.
     *
     * @param context  The Application Context
     * @param customClaims A map of custom claims to be added to the jws payload
     * @param prompt The Prompt to modify the title, subtitle, description
     * @param userKeySelector Collect user key, if not provided [DefaultUserKeySelector] will be used
     * @param deviceAuthenticator A function to return a [DeviceAuthenticator], [deviceAuthenticatorIdentifier] will be used if not provided
     */
    open suspend fun sign(context: Context,
                          customClaims: Map<String, Any> = emptyMap(),
                          prompt: Prompt? = null,
                          userKeySelector: UserKeySelector = DefaultUserKeySelector(),
                          deviceAuthenticator: (type: DeviceBindingAuthenticationType) -> DeviceAuthenticator = deviceAuthenticatorIdentifier) {
        execute(context,
            userKeySelector = userKeySelector,
            deviceAuthenticator = deviceAuthenticator,
            customClaims = customClaims, prompt = prompt)
    }


    /**
     * Sign the challenge with bounded device keys.
     *
     * @param context  The Application Context
     * @param customClaims A map of custom claims to be added to the jws payload
     * @param userKeySelector Collect user key, if not provided [DefaultUserKeySelector] will be used
     * @param deviceAuthenticator A function to return a [DeviceAuthenticator], [deviceAuthenticatorIdentifier] will be used if not provided
     * @param listener The Listener to listen for the result
     * @param prompt The Prompt to modify the title, subtitle, description
     */
    @JvmOverloads
    open fun sign(context: Context,
                  customClaims: Map<String, Any> = emptyMap(),
                  userKeySelector: UserKeySelector = DefaultUserKeySelector(),
                  deviceAuthenticator: (type: DeviceBindingAuthenticationType) -> DeviceAuthenticator = deviceAuthenticatorIdentifier,
                  listener: FRListener<Void?>,
                  prompt: Prompt? = null) {
        val scope = CoroutineScope(Dispatchers.Default)
        scope.launch {
            try {
                sign(context, customClaims, prompt, userKeySelector, deviceAuthenticator)
                Listener.onSuccess(listener, null)
            } catch (e: Exception) {
                Listener.onException(listener, e)
            }
        }
    }

    /**
     * Helper method to execute signing, show biometric prompt.
     *
     * @param context Application Context
     * @param userKey User Information
     * @param deviceAuthenticator A function to return a [DeviceAuthenticator], [getDeviceAuthenticator] will be used if not provided
     * @param customClaims A map of custom claims to be added to the jws payload
     * @param prompt The Prompt to modify the title, subtitle, description
     */
    protected open suspend fun authenticate(context: Context,
                                            userKey: UserKey,
                                            deviceAuthenticator: DeviceAuthenticator,
                                            customClaims: Map<String, Any> = emptyMap(),
                                            prompt: Prompt? = null) {

        deviceAuthenticator.initialize(userKey.userId, prompt?: Prompt(title, subtitle, description))

        if (deviceAuthenticator.isSupported(context).not()) {
            handleException(DeviceBindingException(Unsupported()))
        }

        if (deviceAuthenticator.validateCustomClaims(customClaims).not()) {
            handleException(DeviceBindingException(InvalidCustomClaims()))
            return
        }

        when (val status = deviceAuthenticator.authenticate(context)) {
            is Success -> {
                val jws = deviceAuthenticator.sign(context,
                    userKey,
                    status.privateKey,
                    status.signature,
                    challenge,
                    getExpiration(timeout),
                    customClaims)
                setJws(jws)
            }
            is DeviceBindingErrorStatus -> {
                // All the biometric exception is handled here , it could be Abort or timeout
                handleException(DeviceBindingException(status))
            }
        }

    }

    internal suspend fun execute(context: Context,
                                 userKeyService: UserKeyService = UserDeviceKeyService(context),
                                 userKeySelector: UserKeySelector = DefaultUserKeySelector(),
                                 deviceAuthenticator: (type: DeviceBindingAuthenticationType) -> DeviceAuthenticator,
                                 customClaims: Map<String, Any> = emptyMap(),
                                 prompt: Prompt? = null) {

        try {
            withTimeout(getDuration(timeout)) {
                when (val status = userKeyService.getKeyStatus(userId)) {
                    is NoKeysFound -> handleException(DeviceBindingException(ClientNotRegistered()))
                    is SingleKeyFound -> authenticate(context,
                        status.key,
                        deviceAuthenticator(status.key.authType),
                        customClaims,
                        prompt)
                    else -> {
                        val userKey =
                            userKeySelector.selectUserKey(UserKeys(userKeyService.getAll()))
                        authenticate(context, userKey, deviceAuthenticator(userKey.authType), customClaims)
                    }
                }
            }
        } catch (e: Exception) {
            // This Exception happens only when there is Signing or keypair failed.
            handleException(e)
        }
    }

}